cpp/src/interop/to_arrow_device.cu

```diff
@@ -0,0 +1,697 @@
+    tmp.get(),
+    id_to_arrow_type(
+      column.child(cudf::dictionary_column_view::indices_column_index).type().id())));
+  NANOARROW_RETURN_NOT_OK(ArrowArrayAllocateDictionary(tmp.get()));
+  tmp->length     = column.size();
+  tmp->null_count = column.null_count();
+
+  auto contents = column.release();
+  if (contents.null_mask) {
+    NANOARROW_RETURN_NOT_OK(
+      set_buffer(std::move(contents.null_mask), kValidityBufferIdx, tmp.get()));
+  }
+
+  auto indices_contents =
+    contents.children[cudf::dictionary_column_view::indices_column_index]->release();
+  NANOARROW_RETURN_NOT_OK(
+    set_buffer(std::move(indices_contents.data), kFixedWidthDataBufferIdx, tmp.get()));
+
+  std::unique_ptr<cudf::column>& keys =
+    contents.children[cudf::dictionary_column_view::keys_column_index];
+  NANOARROW_RETURN_NOT_OK(cudf::type_dispatcher(
+    keys->type(), dispatch_to_arrow_device{}, *keys, stream, mr, tmp->dictionary));
+
+  NANOARROW_RETURN_NOT_OK(
+    ArrowArrayFinishBuilding(tmp.get(), NANOARROW_VALIDATION_LEVEL_MINIMAL, nullptr));
+  ArrowArrayMove(tmp.get(), out);
+  return NANOARROW_OK;
+}
+
+struct ArrowDeviceArrayPrivateData {
+  ArrowArray parent;
+  cudaEvent_t sync_event;
+};
+
+void ArrowDeviceArrayRelease(ArrowArray* array)
+{
+  auto private_data = reinterpret_cast<ArrowDeviceArrayPrivateData*>(array->private_data);
+  cudaEventDestroy(private_data->sync_event);
+  ArrowArrayRelease(&private_data->parent);
+  delete private_data;
+  array->release = nullptr;
+}
+
+}  // namespace
+}  // namespace detail
+
+nanoarrow::UniqueSchema to_arrow_schema(cudf::table_view const& input,
+                                        std::vector<column_metadata> const& metadata)
+{
+  CUDF_EXPECTS((metadata.size() == static_cast<std::size_t>(input.num_columns())),
+               "columns' metadata should be equal to the number of columns in table");
+
+  nanoarrow::UniqueSchema result;
+  ArrowSchemaInit(result.get());
+  NANOARROW_THROW_NOT_OK(ArrowSchemaSetTypeStruct(result.get(), input.num_columns()));
+
+  // what i would give for a zip iterator....
+  // thrust::zip_iterator doesn't implement begin/end so I can't use it
+  // for the range-based for loop
+  for (int i = 0; i < input.num_columns(); ++i) {
+    auto child = result->children[i];
+    auto col   = input.column(i);
+    NANOARROW_THROW_NOT_OK(ArrowSchemaSetName(child, metadata[i].name.c_str()));
+    child->flags = col.nullable() ? ARROW_FLAG_NULLABLE : 0;
+
+    if (col.type().id() == cudf::type_id::EMPTY) {
+      NANOARROW_THROW_NOT_OK(ArrowSchemaSetType(child, NANOARROW_TYPE_NA));
+      continue;
+    }
+
+    NANOARROW_THROW_NOT_OK(cudf::type_dispatcher(
+      col.type(), detail::dispatch_to_arrow_type{}, col, metadata[i], *child));
+  }
+
+  return result;
+}
+
+ArrowDeviceArray to_arrow_device(cudf::table&& table,
+                                 rmm::cuda_stream_view stream,
+                                 rmm::mr::device_memory_resource* mr)
```

Comment:
I don't think this function should take an mr. IIUC the only allocations that occur anywhere in this function are of temporaries, such as the bitmask when handling bool columns or the intermediate buffer for decimal types <128 bit. Please correct me if I am wrong, but if not then you should use rmm's current memory resource instead and remove this parameter. See the "Memory Allocation"/"Temporary Memory" section of [the developer guide](https://docs.rapids.ai/api/libcudf/stable/developer_guide).

Reply:
Those aren't temporaries though. The memory allocated for the bitmask for handling bool columns and the buffer for decimal types are what is returned at the end, they aren't used as intermediaries or temporaries.

In the header, this parameter is defaulted to the current memory resource which is what the developer guide says to do

Reply:
Yikes you're totally right. That's my bad, those are definitely outputs. This looks right then, carry on.